src/components/TagManagementDialog/AddTagForm.tsx

```diff
@@ -0,0 +1,214 @@
+import { designTokens } from '../..';
+import type { TFunction } from '../../hooks/useTranslations';
+import { useTranslation } from '../../hooks/useTranslations';
+import { stopKeyDownEvent, withPreventDefault } from './utils';
+import {
+	type SimpleConfirmationDialogProps,
+	SimpleConfirmationDialog,
+} from '../SimpleConfirmationDialog';
+
+const RESERVED_NAMESPACE = 'io.resin.';
+
+const newTagValidationRules = <T extends object>(
+	t: ReturnType<typeof useTranslation>['t'],
+	key: string,
+	existingTags: Array<ResourceTagInfo<T>>,
+) => {
+	return [
+		{
+			test: () => !key || isEmpty(key),
+			message: t('fields_errors.tag_name_cannot_be_empty'),
+		},
+		{
+			test: () => /\s/.test(key),
+			message: t('fields_errors.tag_names_cannot_contain_whitespace'),
+		},
+		{
+			test: () => startsWith(key, RESERVED_NAMESPACE),
+			message: t(`fields_errors.some_tag_keys_are_reserved`, {
+				namespace: RESERVED_NAMESPACE,
+			}),
+		},
+		{
+			test: () =>
+				existingTags.some(
+					(tag) => tag.state !== 'deleted' && tag.tag_key === key,
+				),
+			message: t('fields_errors.tag_with_same_name_exists'),
+		},
+	];
+};
+
+interface AddTagFormProps<T> {
+	t: TFunction;
+	itemType: string;
+	existingTags: Array<ResourceTagInfo<T>>;
+	overridableTags?: Array<ResourceTagInfo<T>>;
+	addTag: (tag: ResourceTagInfo<T>) => void;
+}
+
+export const AddTagForm = <T extends object>({
+	itemType,
+	existingTags,
+	overridableTags = [],
+	addTag,
+}: AddTagFormProps<T>) => {
+	const { t } = useTranslation();
+	const [tagKey, setTagKey] = React.useState('');
+	const [value, setValue] = React.useState('');
+	const [tagKeyIsInvalid, setTagKeyIsInvalid] = React.useState(false);
+	const [error, setError] = React.useState<{ message: string }>();
+	const [canSubmit, setCanSubmit] = React.useState(false);
+	const [confirmationDialogOptions, setConfirmationDialogOptions] =
+		React.useState<SimpleConfirmationDialogProps>();
+
+	const tagKeyInput = React.useRef<HTMLInputElement>(null);
+	const valueInput = React.useRef<HTMLInputElement>(null);
+	const formUuid = `add-tag-form-${uuid.v1()}`;
+
+	const checkNewTagValidity = (key: string) => {
+		const failedRule = newTagValidationRules<T>(t, key, existingTags).find(
+			(rule) => rule.test(),
+		);
+
+		const hasErrors = !!failedRule;
+
+		setTagKeyIsInvalid(hasErrors);
+		setError(failedRule);
+		setCanSubmit(!hasErrors);
+		return hasErrors;
+	};
+
+	const checkTagOverwrites = async () => {
+		const overridableTag = find(overridableTags, {
+			tag_key: tagKey,
+		});
+
+		if (!overridableTag) {
+			return true;
+		}
+		const count = overridableTag.items.length;
+		const result = await new Promise<boolean>((resolve) => {
+			const confirmationOptions = {
+				title: t('warnings.this_would_overwrite_tags', {
+					itemType,
+					count,
+				}),
+				children: (
+					<Typography>
+						{t(`warnings.tag_name_group_exists_and_will_be_overwritten`, {
+							itemType,
+							count,
+						})}
+						<br />
```

Comment:
What is the `<br />` for? Can we just pass the `paragraph` prop to `Typography`?